### DISCUSSION TITLE
how to build a deb package of v0.9.19 for Ubuntu 22.04?

### DISCUSSION THREAD

**Discussion by monkey-jsun** [neutrinolabs/xrdp#2340] [Q&A] [answered]
Currently Ubuntu 22.04 use xrdp v0.9.17 (0.9.17-2ubuntu2).  Unfortunately, this version seems to have [a bug that causes "error loading libvnc.so specified in xrdp.ini"](https://bugs.launchpad.net/ubuntu/+source/xrdp/+bug/1960174), which is fixed in v0.9.18.  There appears to be a critical security fix in v0.9.19.  As such, I like to build v0.9.19 for Ubuntu 22.04.

does anybody know what is the best way to build such a deb package, which hopefully is a drop-in replacement of current v0.9.17 package.  Thanks!

**Comment by matt335672** [comment, marked as answer]
We don't build packages here, so I'm unable to advise you on the best way to go about this. I can give you some pointers however.

The source for the xrdp deb package for Ubuntu and Debian are stored here:-

https://salsa.debian.org/debian-remote-team/xrdp

Their current version on 0.9.19.

If you clone that repo you should be able to use the deb build tool of your choice to build packages from it.

There's information on the Debian Wiki about backporting private packages:-

https://wiki.debian.org/SimpleBackportCreation

that mentions changing the version number of your private backport so that when you update to a version of Ubuntu/Debian with the real 0.9.19, that's what you get.

Hope that's useful.

**Comment by monkey-jsun** [comment]
Below is the script to build the right version I was looking for.

```
#!/bin/bash
  
set -e
set -u
set -x

SCRIPT_DIR="$(dirname "$(readlink -f "$0")")"
ARCH=`uname -m`
if [[ $ARCH == aarch64 ]]; then
    PK_ARCH="arm64"
elif [[ $ARCH == x86_64 ]]; then
    PK_ARCH="amd64"
else
    echo "Unknonw architecture $ARCH"
    exit 1
fi

# step 0 - checking
if [[ $EUID == 0 ]]; then
    echo "This script must NOT be run as root"
    exit 0
fi

# step 1 - uncomment source repo
sudo sed -i -e "s/^# deb-src/deb-src/" /etc/apt/sources.list
sudo apt update

# step 2- install pre-req
sudo apt install -y build-essential fakeroot dpkg-dev
sudo apt build-dep -y xrdp

# step 3 - clone from debian source
git clone https://salsa.debian.org/debian-remote-team/xrdp.git
cd xrdp
git checkout debian/0.9.19-1

# step 4 - build
debian/rules build
fakeroot debian/rules binary
```

**Reply by matt335672** [reply]
That's great - thanks for contributing it.

**Reply by monkey-jsun** [reply]
Just found out a small annoyance.  My script as it is does not apply patches under debian/patches.  Did I miss any step in rebuilding the debian package?

**Reply by matt335672** [reply]
I can't really say - I've never gotten into the complexity of all the build options available for Debian packages.

You could try the following, from https://wiki.debian.org/Diaspora/Packaging/Salsa :-

```
gbp clone --pristine-tar https://salsa.debian.org/debian-remote-team/xrdp
cd xrdp
gbp buildpackage
```

You'll still need your steps to install the dependencies for xrdp. The first command above probably needs a `--debian-branch` switch adding.

**Comment by monkey-jsun** [comment]
It appears adding "dpkg-source --before-build ." before "debian/rules build" in my script solves the problem.
